### DISCUSSION TITLE
Tracing kernel inline functions

### DISCUSSION THREAD

**Discussion by dkogan** [bpftrace/bpftrace#3445] [Development]
Hi.

I'm trying to use bpftrace to probe the kernel in arbitrary locations, and can't figure out how to do it. For instance, I'd like to trace reg_process_hint_user() (at https://github.com/torvalds/linux/blob/d1f2d51b711a3b7f1ae1b46701c769c1d580fa7f/net/wireless/reg.c#L2751) but bpftrace cannot do this because that function is inlined.

Perf appears to be able to do this. On my Debian machine I can `apt install linux-image-6.6.13-amd64-dbg` and then `perf probe -s /home/dima/linux -k /usr/lib/debug/vmlinux-6.6.13-amd64 -m /usr/lib/debug/lib/modules/6.6.13-amd64/kernel/net/wireless/cfg80211.ko --add reg_process_hint_user`.

I think this may be a work-in-progress feature in bpftrace? I see these commits, for instance: https://github.com/bpftrace/bpftrace/pull/3301

I tried that branch; it finds my debug-enabled `vmlinux`, but cannot find the `tcp_connect` symbol inside it, and so decides that bpftrace doesn't have "Kernel DWARF" enabled. This stuff isn't done yet, I guess?

And along the same lines, is there a way to get bpftrace to reference global-but-not-exported symbols? For instance, how do I print the value of `user_alpha2` (https://github.com/torvalds/linux/blob/d1f2d51b711a3b7f1ae1b46701c769c1d580fa7f/net/wireless/reg.c#L288). Perf can do this by (I think) using the DWARF info, so maybe this is a work in progress too?

Thanks!

**Comment by ajor** [comment]
> I think this may be a work-in-progress feature in bpftrace? I see these commits, for instance: https://github.com/bpftrace/bpftrace/pull/3301

Yes, I'm personally very excited by this upcoming feature! It's still a work-in-progress though, as you've found out. It should hopefully make it into the next bpftrace release (0.22).

If you need to trace some inlined functions now, the only way will be to find their offsets manually and then tell bpftrace to kprobe at those offsets. You should be able to get the locations of the inlined functions by loading your kernel's image into a debugger and setting a breakpoint on the function you're interested in, e.g.:
```
$ gdb /boot/vmlinux-6.9.0
(gdb) break reg_pattern_test
Breakpoint 1 at 0xffffffff81bb37b3: reg_pattern_test. (2 locations)
(gdb) info breakpoints 
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   <MULTIPLE>         
1.1                         y   0xffffffff81bb37b3 in reg_pattern_test at drivers/net/ethernet/intel/igb/igb_ethtool.c:1221
1.2                         y   0xffffffff81bb38e3 in reg_pattern_test at drivers/net/ethernet/intel/ixgbe/ixgbe_ethtool.c:1552
```

Then probe at the addresses of each inlined instance:
```
kprobe:0xffffffff81bb37b3,
kprobe:0xffffffff81bb38e3
{
  ...
}
```

**Comment by ajor** [comment]
> And along the same lines, is there a way to get bpftrace to reference global-but-not-exported symbols?

That's a very nice idea. I've opened an issue for it here: #3446.

It's not something that's being worked on right now - any help would be appreciated!

bpftrace now incorporates liblldb, so as long as the LLDB command line tool can access these variable addresses, it should just be a matter of finding the right LLDB API call to make to make that info available in bpftrace.

**Comment by ttreyer** [comment]
> I tried that branch; it finds my debug-enabled `vmlinux`, but cannot find the `tcp_connect` symbol inside it, and so decides that bpftrace doesn't have "Kernel DWARF" enabled. This stuff isn't done yet, I guess?

Hi @dkogan, thanks for reporting the issue about the "Kernel DWARF" feature detection.
I need to find a better way to detect if the Kernel has DWARF or not.

Otherwise, the branch should be pretty close to ready.
If you'd like to experiment with it, you can try changing `"tcp_connect"` by another function, or just bypass the test by returning `true`.
